### PR TITLE
Modernize testing

### LINKED ISSUE #16: Fix atomic density path if missing final `/`
Chargemol has a nasty bug that make it crash if the path to the atomic densities does not ends with a `/`, for example:
```
    ddec_params = Dict(
        dict={
            'net charge': 0.0,
            'charge type': 'DDEC6',
            'periodicity along A, B, and C vectors': [True, True, True],
            'compute BOs': False,
            'atomic densities directory complete path': 'path/to/atomic_densities', #WILL FAIL! 
            'input filename': 'valence_density',
        })
```
Since this is a parameter that the user needs to provide every time, depending on the computer, make the plugin fix this problem by adding a `/` if missing.

### PR DESCRIPTION
**Pull request by ltalirz**
fix #16 

This PR

 * moves tests from docker to `aiida-testing`. In particular, it is now inconsequential how long the simulation codes need to run, since the mock executable spits out results instantaneously
 * starts validating input dictionary of `aiida-ddec`
 * allows specifying the atomic densities directory via a `DDEC_ATOMIC_DENSITIES_DIRECTORY` extra on the ddec Code instance
 * makes sure that the atomic densities directory ends with `/`

### CODE REVIEW

**Comment by ltalirz** [comment]
Note: The non-portable input files currently are a challenge for the mock executable of `aiida-testing` and I've opened an issue for this https://github.com/aiidateam/aiida-testing/issues/48

Until this is resolved, one will simply need to provide the same path to the atomic densities directory for the CI tests as one has used locally to generate the test data (as documented in the README).

**Comment by ltalirz** [comment]
Tests pass: https://github.com/ltalirz/aiida-ddec/actions/runs/375011721

**Review by yakutovicha — COMMENTED** [review]
Thanks, @ltalirz. I have a few questions, comments:

1. I would appreciate a lot if we could stick to one-pr-one-change philosophy. Of course, there are cases where multiple changes are unavoidable or too trivial, but I guess this PR isn't the case. To clarify my point I would like to emphasize the following:
    - Small and single-objective PRs are easier to test, to understand and to verify. 
    - If there is one thing that we disagree on, it could result in a delay of merging other things that are addressed in this PR.
 
2. I am not a big expert of the mock executables, but I don't understand why do we need to include the atomic densities (~50 files) and the `aiida-RESTART.wfn` binary file. Can they be removed?

3. I hoped to make the 1.0.0 with the support of python2. I won't stand in your way and block the PR because of that. However, if this is not hard, I would appreciate keeping python2 support until 1.1 release. I do understand, however, that python2 is probably not used by anyone anymore, so in the reality, this is most probably not the issue. Therefore, I leave it up to you.

4. When you merge this PR please squash the related topics into one commit and rebase-merge the PR afterwards.

5. I looked at the code, it all seems to be fine, but I haven't tested it because seeing the complexity of the PR, that would probably take a lot of time.

**Comment by ltalirz** [comment]
Thanks @yakutovicha for the review!

> 1. I would appreciate a lot if in the future we could stick to one-pr-one-change philosophy. Of course, there are cases where multiple changes are unavoidable or too trivial, but I guess this PR isn't the case. To clarify my point I would like to emphasize the following:
>    
>    * Small and single-objective PRs are easier to test, to understand and to verify.
>    * If there is one thing that we disagree on, it could result in a delay of merging other things that are addressed in this PR.
You're right... I needed some of the changes in this branch for aiida-lsmo, that's why I ended up doing it in one go, but it's not a good practice and I will do it better in the future.


> 2. I am not a big expert of the mock executables, but I don't understand why do we need to include the atomic densities (~50 files) 

The atomic densities make it possible to run the tests without having to tell the code where to find the atomic densities directory (via an environment variable). I've already limited them to the elements used in the examples, so they're just 284kB overall.
You could remove them, but then it means you will always need to export an environment variable before running the tests...

> and the `aiida-RESTART.wfn` binary file. Can they be removed?

I was wondering about this as well - do any of the workchains here restart calculations (the same question applies to `aiida-lsmo`)?
I wasn't sure, so I went with the sage route, but if that is not the case, then I will definitely remove it.

> 3. I hoped to make the 1.0.0 with the support of python2. I won't stand in your way and block the PR because of that. However, if this is not hard, I would appreciate keeping python2 support until 1.1 release. I do understand, however, that python2 is probably not used by anyone anymore, so in the reality, this is most probably not the issue. Therefore, I leave it up to you.

Ok, no problem, I will put the hooks back.

> 4. When you merge this PR please squash the related topics into one commit and rebase-merge the PR afterwards.

Ok.

File: .pre-commit-config.yaml
Comment (ltalirz) [reply .pre-commit-config.yaml:16]: I've put it back

File: README.md
Comment (ltalirz) [reply README.md:13]: will do!

**Comment by ltalirz** [comment]
@yakutovicha Incorporated your suggestions and tests pass again https://github.com/ltalirz/aiida-ddec/actions/runs/381475493

Please have a look at my reply to point 2. above; the others should have been addressed

**Review by yakutovicha — COMMENTED** [review]
@ltalirz, I found one little issue (see below).

> The atomic densities make it possible to run the tests without having to tell the code where to find the atomic densities directory (via an environment variable).

But if you are running a mock executable, why does it require the atomic densities?

Again, I am asking this because I probably don't fully understand the concept of the mock executable. To me, if you are running a fake (mock) executable - it should not rely on any files apart from the "standard" inputs.

**Comment by yakutovicha** [comment]
> I was wondering about this as well - do any of the workchains here restart calculations (the same question applies to `aiida-lsmo`)?

Not sure about `aiida-lsmo`, but here you *can* provide previous calculation as the `parent_calc` input. However, this is not strictly necessary.

**Comment by ltalirz** [comment]
> But if you are running a mock executable, why does it require the atomic densities?
> 
> Again, I am asking this because I probably don't fully understand the concept of the mock executable. To me, if you are running a fake (mock) executable - it should not rely on any files apart from the "standard" inputs.

Right, my comment was not clear - you *can* run the tests without needing the atomic densities folder (as long as you don't change them).
My comment concerned the case where you would like to regenerate test data, e.g. if one of the inputs changes.

**Comment by yakutovicha** [comment]
> Right, my comment was confusing - you _can_ run the tests without needing the atomic densities folder (as long as you don't change the calculation inputs).
> My comment was aimed at the case where you would like to regenerate test data, e.g. if one of the inputs changes.

Ok, so it is clear now. My understanding is that to generate the apps one needs to install the ddec code, which itself comes with the atomic densities. Therefore (if I would be doing it) I would add one more line to the readme that one should add atomic densities path to the system variable.

> > Not sure about `aiida-lsmo`, but here you _can_ provide previous calculation as the `parent_calc` input. However, this is not strictly necessary.
> 
> The question that matters is whether this is being used in the tests.
> If not, we can remove the .wfn (but in the end it's just a 4KB file).

No, for the tests this isn't needed.

**Comment by ltalirz** [comment]
> Ok, so it is clear now. My understanding is that to generate the apps one anyways needs to install the ddec code, which itself comes with the atomic densities. Therefore (if I would be doing it) I would add one more line to the readme that one should add atomic densities path to the system variable.

I looked into making this change, but then realized that it is not ideal/straightforward (the hashes of the code inputs would not fit due to the different file paths in the input file).

The root of the problem is that by forcing users to put the directory path into the input file, the atomic densities directory effectively becomes part of the input.
I propose that I will 
  1. Keep the reduced atomic_densities dir in the test data
  2. Modify the calculation plugin such that it creates a symlink to that directory

This way, the input directory determines the outcome of the calculation, and the tests should be easy to update if necessary.

**Comment by ltalirz** [comment]
@yakutovicha can I go ahead with rebase & merge?

**Review by yakutovicha — APPROVED** [review]
thanks @ltalirz, looks good to me. Please go ahead with merging.

**Comment by ltalirz** [comment]
Thanks!

**Pull request merged**
